File: proj1/python/matrixprod.py
#!/usr/bin/env python3

from pypapi import papi_high
from pypapi import events as papi_events

from time import time
import sys


def papiStartCounters():
    papi_high.start_counters([
        papi_events.PAPI_L1_DCM,
        papi_events.PAPI_L2_DCM,
    ])


def papiStopCounters():
    return papi_high.stop_counters()


def multiplyMatrix(mat_size, is_line=False):
    mat1 = [1.0] * (mat_size * mat_size)
    mat2 = [1.0 + i for i in range(mat_size) for _ in range(mat_size)]
    mat_res = [0.0] * (mat_size * mat_size)

    start_time = time()
    if is_line:
        for i in range(mat_size):
            for k in range(mat_size):
                for j in range(mat_size):
                    mat_res[i*mat_size + j] += mat1[i*mat_size + k] * mat2[k*mat_size + j]
    else:
        for i in range(mat_size):
            for j in range(mat_size):
                acc = 0
                for k in range(mat_size):
                    acc += mat1[i*mat_size + k] * mat2[k*mat_size + j]
                mat_res[i*mat_size + j] = acc

    end_time = time()
    return '{:3.3f}'.format(end_time - start_time)

def printUsage(path):
    print(f'usage: {path} <algorithm> <matrix_size>')
    print('algorithms:')
    print('\t1: regular')
    print('\t2: line by line')

if __name__ == "__main__":
    if len(sys.argv) < 3:
        printUsage(sys.argv[0])
        sys.exit(1)

    algorithm = int(sys.argv[1])
    mat_size = int(sys.argv[2])

    if algorithm < 1 or algorithm > 3:
        print("Invalid algorithm!")
        printUsage(sys.argv[0])
        exit(2)

    papiStartCounters()

    if algorithm == 1:
        time = multiplyMatrix(mat_size)
    elif algorithm == 2:
        time = multiplyMatrix(mat_size, True)      

    [l1_dcm, l2_dcm] = papiStopCounters()

    print(f'{time} {l1_dcm} {l2_dcm}')

File: proj1/java/MatrixProd.java
import java.util.Arrays;

class MatrixProd {
    private static enum Algorithm {
        REGULAR, 
        LINE_BY_LINE,
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            printUsage();
            System.exit(1);
        }

        int algorithm = Integer.parseInt(args[0]);
        int mat_size = Integer.parseInt(args[1]);

        if (algorithm < 1 || algorithm > Algorithm.values().length) {
            System.out.println("Invalid algorithm!");
            printUsage();
            System.exit(2);
        }

        String exec_time = multiplyMatrix(mat_size, algorithm == 1 ? Algorithm.REGULAR : Algorithm.LINE_BY_LINE);
        System.out.println(String.format("%s N/A N/A", exec_time));
    }

    private static String multiplyMatrix(int mat_size, Algorithm algorithm) {
        double[] mat1 = new double[mat_size * mat_size];
        double[] mat2 = new double[mat_size * mat_size];
        double[] mat_res = new double[mat_size * mat_size];

        Arrays.fill(mat1, 1.0);
        for (int i = 0; i < mat_size; ++i) {
            Arrays.fill(mat2, mat_size * i, mat_size * (i + 1), 1.0 + i);
        }
        Arrays.fill(mat_res, 0.0);

        long start_time = System.nanoTime();

        if (algorithm == Algorithm.LINE_BY_LINE) {
            for (int i = 0; i < mat_size; ++i) {
                for (int k = 0; k < mat_size; ++k) {
                    for (int j = 0; j < mat_size; ++j) {
                        mat_res[i*mat_size + j] += mat1[i*mat_size + k] * mat2[k*mat_size + j];
                    }
                }
            }
        } else {
            for (int i = 0; i < mat_size; ++i) {
                for (int j = 0; j < mat_size; ++j) {
                    double acc = 0;
                    for (int k = 0; k < mat_size; ++k) {
                        acc += mat1[i*mat_size + k] * mat2[k*mat_size + j];
                    }
                    mat_res[i*mat_size + j] = acc;
                }
            }
        }

        long end_time = System.nanoTime();

        return String.format("%.3f", (double)((end_time - start_time)/1e9));
    }

    private static void printUsage() {
        System.out.println("usage: MatrixProd <algorith> <matrix size>");
        System.out.println("algorithms:");
        System.out.println("\t1: regular");
        System.out.println("\t2: line by line");
    }
}